Issue #8186: add new check NoWhiteSpaceBeforeCaseDefaultColonCheck

AkMo3 opened the pull request:
Closes #8186
    

- New Check NoWhiteSpaceBeforeCaseDefaultColonCheck

- This is continutation of PR #8187 by @wilcoln.
- Contribution PR : checkstyle/contribution#548

New module config: https://github.com/AkMo3/akmo3.github.io/raw/master/ColonCheck/WhiteSpaceBeforeColon.xml
New Report: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8cce311_2021034422/reports/diff/index.html
___
Old Report: https://akmo3.github.io/NoWhiteSpace/reports/diff/index.html

nmancus1 commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,106 @@
+/*
+* Config = default
+*/
+public class InputNoWhitespaceBeforeCaseDefaultColon {
```
It would be a good idea to some switch expressions in here, you can find some more challenging examples in https://github.com/checkstyle/checkstyle/blob/master/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java

AkMo3 replied:
Done, please see recent changes and suggest any changes if required.

nmancus1 replied:
I do not see any switch statements/ expressions with compound/ block statements, it would be good to add a few of them.

AkMo3 replied:
Do mean nested switch kind of statements, I didn't get what you mean by compound/block statement?

nmancus1 replied:
Like this:

```java
        switch(x) {
            case 1: {  // <- curly brace here, now this is a "block statement"
                // do stuff
                break;
            }
            default: {  // <- curly brace here, now this is a "block statement"
                // do other stuff
            }
        }
```

AkMo3 replied:
Done. Sir, can you just please help me on you to write PR description for Travis to pass?

nmancus1 replied:
Travis will not pass until  https://github.com/checkstyle/contribution/pull/471 is merged.

AkMo3 replied:
All right thanks, Please check the updated examples. I will fix the CI.

nmancus1 replied:
Resolved.

nmancus1 commented:
@AkMo3 you can keep noise down in the reports by adding this module:
```xml
    <module name="SuppressionSingleFilter">
        <property name="message" value="Exception occurred while parsing"/>
        <property name="checks" value="Checker"/>
    </module>
```

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,189 @@
+ *             break;
+ *     }
+ *
+ *     switch(2) {
+ *         case 2: // OK
+ *             break;
+ *         default: // OK
+ *             break;
```
I think it would be good to show a line wrapped example here, like you have done in your test input.

AkMo3 replied:
Done.

nmancus1 replied:
All we really need to do is:

```java
         case 2: // OK
             break;
         default
                      : //  violation
```
Also, I don't think we need two different examples for this check. We can probably just merge these two examples.

AkMo3 replied:
Done

AkMo3 replied:
@nmancus1 Is it resolved? Please comment on others too if they are resolved.

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.xml:
```diff
@@ -0,0 +1,16 @@
+             name="NoWhitespaceBeforeCaseDefaultColon"
+             parent="com.puppycrawl.tools.checkstyle.TreeWalker">
+         <description>&lt;p&gt;
+ Checks that there is no whitespace before a colon token of a switch case or default.
```
Just a thought:
`Checks that there is no whitespace before the colon in a switch label.`?
switch case or default == case label https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.11

AkMo3 replied:
Done

nmancus1 replied:
Resolved.

nmancus1 reviewed (changes requested):
Few items:

nmancus1 commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColonEnumAndStrings.java:
```diff
@@ -0,0 +1,81 @@
+        };
+    }
+
+    int assignement(Day day) {
```
I wonder why spellcheck did not catch this.

AkMo3 replied:
Resources are not checked by spell check, so this does not raise any violation.

https://github.com/checkstyle/checkstyle/blob/b15820a77a1cc27c3ad247602f4c89f442d22bdb/.ci/jsoref-spellchecker/exclude.pl#L11

nmancus1 reviewed (commented):
One more thing:

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,194 @@
+     */
+    private boolean checkWhitespaceBeforeColonOfCase(DetailAST colonAst) {
+        final DetailAST previousSibling = colonAst.getPreviousSibling();
+        final boolean areOnSameLine = TokenUtil.areOnSameLine(previousSibling, colonAst);
+        boolean horizontalWhitespaceDetected = false;
+        if (areOnSameLine) {
+            int offset = 0;
+            if (previousSibling.getType() == TokenTypes.BLOCK_COMMENT_BEGIN) {
+                offset = 1;
+            }
+            horizontalWhitespaceDetected =
+                    colonAst.getColumnNo() != getLastColumnNumberOf(previousSibling) + offset;
+        }
+        else {
+            log(colonAst, MSG_KEY_ON_SEPARATE);
+        }
+        return horizontalWhitespaceDetected;
```
I think that the duplication of code for checking if the tokens are on different lines is unnecessary; can we extract this and make a new method just for checking across lines? So we have  `checkWhitespaceBeforeColonOfDefault()`, `checkWhitespaceBeforeColonOfCase()`, then maybe `checkIfColonIsSameLine()` or something similar, all of which are called from `visitToken()`?.

AkMo3 replied:
This definitely makes the code clean, but we cannot use here since we comparing different tokens with `colonAst` for same line. I mean we can do but then there would be some more extra conditions to check like if something is not null and all.

nmancus1 replied:
My thoughts were something like this in `visitToken()`:

```java
        if (!isColonOnSameLine) {
            log();
        } 
        else if (isWhitespacePrecedingColonOfDefault() 
              || isWhitespacePrecedingColonOfCase()) {
           log();
       }
```

I think it is a bit confusing to have a boolean method that also logs violations.  We should at least specify in the javadocs for that these methods also log violations for tokens on different lines, but I think it would be best to keep this code separate.  @romani @strkkk @pbludov please share your thoughts on this and below suggestions.

nmancus1 replied:
@romani @strkkk @pbludov ping

strkkk replied:
I do not like current way. It seems to me that we can log violation in method and then log it one more time after we returned something from method.
It is better to log it in one place as @nmancus1 suggested

AkMo3 replied:
Done.

AkMo3 replied:
@nmancus1 @strkkk Semaphore is failing on `ci/no-exception-test.sh guava-with-google-checks`, Error log says:
```
Error rendering Maven report: Failed during checkstyle configuration: cannot initialize module TreeWalker - cannot initialize
module NoWhitespaceBeforeCaseDefaultColon - Unable to instantiate 'NoWhitespaceBeforeCaseDefaultColon'
class, it is also not possible to instantiate it as 
com.github.sevntu.checkstyle.checks.annotation.NoWhitespaceBeforeCaseDefaultColon,
...
```
Why should sevntu check config being used? Is it introduced in #9278?

nmancus1 replied:
> Why should sevntu check config being used? 

This is just an error message from the  mvn-checkstyle-plugin.

> Is it introduced in #9278?

The problem is that while diff.groovy doesn't use the branch specified in single mode, it still checks it out. So, this CI job ran on master instead of your PR branch. PR to fix this is at https://github.com/checkstyle/checkstyle/pull/9868. Once that is merged, rebase this PR.

AkMo3 replied:
Ok, thanks.

nmancus1 replied:
Thanks for catching this. The reason that we are trying to stop using launch.groovy is that it  is deprecated and will be removed  soon.

romani replied:
@AkMo3 , please rebase

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,194 @@
+     * Checks if there is a whitespace before the colon of a switch default.
+     *
+     * @param colonAst DetailAST to check.
+     * @return whether there is a whitespace before colonAst or not.
```
`@return true if there is whitespace preceding colon ast` this removes any ambiguity. Same for above javadoc.

AkMo3 replied:
Done.

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,194 @@
+     * @param colonAst DetailAST to check.
+     * @return whether there is a whitespace before colonAst or not.
+     */
+    private boolean checkWhitespaceBeforeColonOfDefault(DetailAST colonAst) {
```
What do you think about `isWhitespacePrecedingColonOfDefault()`?

AkMo3 replied:
Sir, I would not prefer to change these names as these names where already approved before so lets keep them as it is.

AkMo3 replied:
Sir, I would not prefer to change these names as they were already approved by members before.

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,194 @@
+     * @param colonAst DetailAST to check.
+     * @return whether there is a whitespace before colonAst or not.
+     */
+    private boolean checkWhitespaceBeforeColonOfCase(DetailAST colonAst) {
```
What do you think about `isWhitespacePrecedingColonOfCase()`?

AkMo3 replied:
Please sir above

nmancus1 replied:
Resolved.

nmancus1 reviewed (changes requested):
Few more items:

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,205 @@
+     * @param isColonOfDefault boolean parameter representing if the colon is after default.
+     * @return true, if colon case is in different line as of case or default.
+     */
+    private static boolean isColonOnDifferentLine(DetailAST colonAst, boolean isColonOfCase,
+                                      boolean isColonOfDefault) {
+        boolean sameLine = true;
+        if (isColonOfCase) {
+            final DetailAST previousSibling = colonAst.getPreviousSibling();
+            sameLine = TokenUtil.areOnSameLine(previousSibling, colonAst);
+        }
+        else if (isColonOfDefault) {
+            final DetailAST literalDefault = colonAst.getParent();
+            sameLine = TokenUtil.areOnSameLine(literalDefault, colonAst);
+        }
+        return !sameLine;
+    }
```
Can we extract `previousSibling` and `literalDefault` to just have a `previousToken`? Then we can just return `!TokenUtil.areOnSameLine(previousToken, colonAst);` I would also just pass in the parent token that you have already declared in `visitToken` and check for it's type in the `if`/ `else if` conditions, to eliminate the boolean parameters.

Something like:
```java
    private static boolean isColonOnDifferentLine(DetailAST colonAst, DetailAst parent) {
        final DetailAst previousToken;
        if (parent.getType() == TokenTypes.LITERAL_CASE) {
            previousToken = colonAst.getPreviousSibling();
        }
        else if (parent.getType() == TokenTypes.LITERAL_DEFAULT) {
            previousToken = colonAst.getParent();
        }
        return !TokenUtil.areOnSameLine(previousToken, colonAst);
     }

```

AkMo3 replied:
Done.

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,205 @@
+        final DetailAST parent = ast.getParent();
+        final boolean isColonOfCase = parent.getType() == TokenTypes.LITERAL_CASE;
+        final boolean isColonOfDefault = parent.getType() == TokenTypes.LITERAL_DEFAULT;
+        final boolean violation = isColonOfCase && checkWhitespaceBeforeColonOfCase(ast)
+                || isColonOfDefault && checkWhitespaceBeforeColonOfDefault(ast);
```
If you use the below suggestion, you can just incorporate this boolean into the `else if` branch below.

AkMo3 replied:
This is not clear, can you please explain in bit detail

nmancus1 replied:
On second thought, disregard that comment. 

What do you think about naming `violation` something like `isWhiteSpaceBeforeColon`? I think that would read well:

```java
        if (isColonOnDifferentLine(ast, parent)) {
            log(ast, MSG_KEY_ON_SEPARATE);
        }
        else if (isWhiteSpaceBeforeColon) {
            log(ast, MSG_KEY, ast.getText());
        }
```

This could even be a method itself (`isWhiteSpaceBeforeColon()`), and simply `return  isColonOfCase && checkWhitespaceBeforeColonOfCase(ast)
                || isColonOfDefault && checkWhitespaceBeforeColonOfDefault(ast);`, giving us:
                

```java
        if (isColonOnDifferentLine(ast, parent)) {
            log(ast, MSG_KEY_ON_SEPARATE);
        }
        else if (isWhiteSpaceBeforeColon(ast)) {
            log(ast, MSG_KEY, ast.getText());
        }
```

AkMo3 replied:
> What do you think about naming violation something like isWhiteSpaceBeforeColon? I think that would read well:

Totally in its favor. Done.

nmancus1 replied:
Resolved.

nmancus1 reviewed (changes requested):
A few more thoughts, this is looking really good now:

nmancus1 reviewed (approved):
Looks great. @strkkk please assign next reviewer.

strkkk reviewed (commented):
@AkMo3 please generate new report (you can use github action for it)

AkMo3 commented:
GitHub, generate report

AkMo3 commented:
GitHub, generate report

AkMo3 commented:
GitHub, generate report

AkMo3 commented:
@strkkk Report generation is constantly failing. Can you please confirm it is a bug after recent changes or there is some mistake from my side. Error log:
```
Resetting git sources to commit 'master'
Running command: git reset --hard master
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

strkkk commented:
I see it is failing on spring-integration. Please try to exclude it from list and retry.

nmancus1 commented:
@AkMo3 please use default projects file (unspecified). Spring changed master branch to main two weeks ago.

AkMo3 commented:
GitHub, generate report

strkkk reviewed (approved):
@AkMo3 nice work :)

AkMo3 commented:
@romani Please start review

electrum commented:
Can you add some tests for [switch expressions](https://openjdk.java.net/jeps/361) as well?

AkMo3 commented:
I had already added expressions https://github.com/checkstyle/checkstyle/blob/d9643d995d978ac5bcc5dd052a2fa985a9ad6597/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColonEnumAndStrings.java#L19-L24

Extra test cases were not added as these were just skipped these do not have "`:` (TokenTypes.COLON)" which is the required token for this check.

Let me know if you want more test cases.

electrum commented:
Thanks, I missed that one. Looks good to me!

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,210 @@
+
+/**
+ * <p>
+ * Checks that there is no whitespace before the colon in a switch label.
```
`Checks that there is no whitespace before the colon in a switch block.`

https://docs.oracle.com/javase/specs/jls/se16/html/jls-14.html#jls-14.11

AkMo3 replied:
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,210 @@
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (isColonOnDifferentLine(ast)) {
```
first condition should be for `isInSwitch(ast)` that validate that colon is for `case` or `default`. 
to make sure there is no execution of Ckeck at all on `for(int i: vallues)`, please add it to test cases.

romani replied:
We can write a lot of code that will return required results. But better to never let code execute on non target structure.
Please filterout all at the beginning.

AkMo3 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,210 @@
+        final DetailAST parent = colonAst.getParent();
+        final boolean isColonOfCase = parent.getType() == TokenTypes.LITERAL_CASE;
+        final boolean isColonOfDefault = parent.getType() == TokenTypes.LITERAL_DEFAULT;
+        return isColonOfCase && checkWhitespaceBeforeColonOfCase(colonAst)
```
checkWhitespaceBeforeColonOfCase ==> isWhitespaceBeforeColonOfCase

`check` might mean return more then `true/false`

AkMo3 replied:
Done

romani reviewed (changes requested):
items:

AkMo3 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,210 @@
+ *         default
+ *               : // violation, ':' should not be in separate line
+ *             break;
+ *     }
```
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,227 @@
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (isInSwitch(ast)) {
+            if (isColonOnDifferentLine(ast)) {
```
Please point me discussion where we made decision  to have separate message on this.
Separate line already mean there is space before `:`

AkMo3 replied:
> Please point me discussion where we made decision to have separate message on this.

We have not discussed it.

> Separate line already mean there is space before `:`

True, but this statement becomes confusing for some cases like 
```
case 1
      :  break;
```
and 
```
case 1
: break;
```

It both cases if we give message `: is preceded with whitespace.`, doesn't gives a clear picture.

romani replied:
We need to compare line and column.

>It both cases if we give message : is preceded with whitespace., doesn't gives a clear picture.

Hm, for me it is clear. 
There might be tab symbol in between (not a whitespace) and this might be confusing a bit. But can change wording a bit to be general on any non printable symbol in between.

AkMo3 replied:
> We need to compare line and column.

I will do change.

> There might be tab symbol in between (not a whitespace) and this might be confusing a bit. But can change wording a bit to be general on any non printable symbol in between.

What do think of a single message like : `''{0}'' is preceded with non printable symbol.` for both cases?

AkMo3 replied:
Is the name of the check and its definition is required to be changed? 
I mean should we change name from `NoWhitespaceBeforeCaseDefaultColonCheck` ->`NoNonPrintableSymbolBeforeCaseDefaultColonCheck`
and the definition of check to 
`Checks that there is no whitespace before the colon in a switch block.` -> `Checks that there is no non printable symbol before the colon in a switch block.`

romani replied:
>is preceded with whitespace

Let's use this, us us simple and good enough.
And keep Check name.

AkMo3 replied:
Done.

romani reviewed (approved):
Items:

AkMo3 commented:
GitHub, generate report

AkMo3 commented:
GitHub, generate report

AkMo3 commented:
Identical failure in both report generation
```
 Get:24 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 groovy all 2.4.17-4ubuntu1 [12.9 MB]
Err:23 http://security.ubuntu.com/ubuntu focal-updates/universe amd64 libxstream-java all 1.4.11.1-1ubuntu0.1
  404  Not Found [IP: 52.252.75.106 80]
Get:25 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 libjcsp-java all 1.1-rc4-2 [607 kB]
Get:26 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 libjsr166y-java all 1.7.0-2 [599 kB]
Get:27 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 libmultiverse-core-java all 0.7.0-4 [219 kB]
Get:28 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 libgpars-groovy-java all 1.2.1-10 [1116 kB]
Get:29 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 libjcommander-java all 1.71-3 [69.3 kB]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/universe/libx/libxstream-java/libxstream-java_1.4.11.1-1ubuntu0.1_all.deb  404  Not Found [IP: 52.252.75.106 80]
```

Manually generated new report: https://akmo3.github.io/NoWhiteSpace/reports/diff/index.html

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,214 @@
+        final DetailAST parent = colonAst.getParent();
+        final boolean isColonOfCase = parent.getType() == TokenTypes.LITERAL_CASE;
+        final boolean isColonOfDefault = parent.getType() == TokenTypes.LITERAL_DEFAULT;
+        return isColonOfCase && isWhitespaceBeforeColonOfCase(colonAst)
```
make it like:
```
if (parent.getType() == TokenTypes.LITERAL_CASE)
{
 result  = ...
} else {
 // it is default 
 result = ...
}
```

AkMo3 replied:
Done.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,214 @@
+            horizontalWhitespace =
+                    colonAst.getColumnNo() != getLastColumnNumberOf(commentBlock) + 1;
+        }
+        return horizontalWhitespace;
```
make it `return result;`
name of variable is too confusing.

AkMo3 replied:
Done.
Yes, it was confusion. I didn't change it since it was not pointed in earlier PR.

https://github.com/wilcoln/checkstyle/blob/c5fd71c27ccdb7485c08853517019e60484a15db/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java#L135

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,214 @@
+     * @param colonAst DetailAST to check.
+     * @return true, if colon case is in different line as of case or default.
+     */
+    private static boolean isColonOnDifferentLine(DetailAST colonAst) {
```
isOnDifferentLineWithPreviosToken

AkMo3 replied:
Done

romani reviewed (changes requested):
items:

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,214 @@
+     */
+    private static boolean isWhitespaceBeforeColonOfDefault(DetailAST colonAst) {
+        final boolean horizontalWhitespace;
+        final DetailAST commentBlock = colonAst.getPreviousSibling();
```
why colon previous sibling is comment block ?
please change names or ....

AkMo3 replied:
Because between a `DEFAULT` and `COLON` only comment can be present.
Now changed to `previousSibling`.

romani replied:
if it can be present there should be code like:
```
final DetailAST ast = colonAst.getPreviousSibling();
if ( ast  == COMMENT_BLOCK) { .... 
```

AkMo3 replied:
It can be present as shown in this example https://github.com/checkstyle/checkstyle/blob/bfe149624e6a9003c2f7bef163c4c8bea39bca7b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java#L87-L88

But, we do not care if it is a comment or any other token, all we care about is its length. So we should be good with current change also.

romani reviewed (changes requested):
items:

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,220 @@
+    private static boolean isWhitespaceBeforeColonOfDefault(DetailAST colonAst) {
+        final boolean result;
+        final DetailAST previousSibling = colonAst.getPreviousSibling();
+        if (previousSibling == null) {
```
why in default processing you do not have code like
`if (previousSibling.getType() == TokenTypes.BLOCK_COMMENT_BEGIN) {` ?

AkMo3 replied:
Please see https://github.com/checkstyle/checkstyle/pull/9728#discussion_r633028187

romani replied:
yes it is possible to have comments.
But why it is different between `case:` and `default:` ?
they are case by structure 

`colonAst.getColumnNo() != getLastColumnNumberOf(previousSibling) + offset;`
vs
`colonAst.getColumnNo() != getLastColumnNumberOf(previousSibling) + 1;`

AkMo3 replied:
There is a difference because for `LITERAL_CASE` we always have an `EXPR` before the colon, for which the `offset` is 0.
When there is `BLOCK_COMMENT`, we make offset to be 1 otherwise it is 0 as shown in the below code.

https://github.com/checkstyle/checkstyle/blob/bfe149624e6a9003c2f7bef163c4c8bea39bca7b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java#L160-L164

Now, for `LITERAL_DEFAULT` we are sure that there can only be `BLOCK_COMMENT`, nothing else can be used there. So we are directly adding 1, whenever there is a previous sibling. 

Hence, we are treating `BLOCK_COMMENT` same in both cases i.e. we add an `offset` of 1 in both cases.

romani replied:
ok now I see it
```
|--CASE_GROUP -> CASE_GROUP [5:12]
|   |--LITERAL_CASE -> case [5:12]
|   |   |--EXPR -> EXPR [5:17]
|   |   |   `--NUM_INT -> 1 [5:17]
|   |   `--COLON -> : [5:18]
|   `--SLIST -> SLIST [6:16]
|       `--LITERAL_BREAK -> break [6:16]
|           `--SEMI -> ; [6:21]
|--CASE_GROUP -> CASE_GROUP [7:12]
|   |--LITERAL_DEFAULT -> default [7:12]
|   |   `--COLON -> : [7:20]
```

romani reviewed (changes requested):
items:

romani commented on src/xdocs/config_whitespace.xml:
```diff
@@ -1687,6 +1687,93 @@ Lists.charactersOf("foo")
       </subsection>
     </section>
+    <section name="NoWhitespaceBeforeCaseDefaultColon">
+      <p>Since Checkstyle 8.43</p>
```
please bump version here and in javadoc.

AkMo3 replied:
Done

romani reviewed (changes requested):
last minor item:

romani reviewed (approved):
@AkMo3 , thanks a lot for your hard work to make it happen

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,220 @@
+     * @param colonAst DetailAST to check.
+     * @return true, if there is whitespace preceding colonAst.
+     */
+    private static boolean isWhitespaceBeforeColonOfDefault(DetailAST colonAst) {
```
`isWhitespaceBeforeColonOfCase` has custom support for comments. How come this isn't the same for default?

AkMo3 replied:
Default also has comment support, only thing for `case` we need to specifically handle comment as there can be other expressions, but in case of `default`, since there can be only comment we are simply checking if there is any other element before colon
https://github.com/checkstyle/checkstyle/blob/8cce31164d76c5b4c527a163f6d94a3b4805a316/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java#L176-L177

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * Checks if there is a whitespace before the colon of a switch case.
```
2nd sentence should mention comment support.

AkMo3 replied:
I think this is not required as comment handling is basic requirement it is being done in both `case` and `default`

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * Checks if the colon is on same line as of case or default.
```
> same line

Method is named "different" line. I would think the summary should follow the same logic and not go towards the negative of what we are doing.

AkMo3 replied:
Done

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * Checks if there is a whitespace before the colon of a switch default.
```
> colon of a switch default

Truthfully we are checking against the previous token. There is no check or guarantee that it is the colon or switch, right?

AkMo3 replied:
We are sure as we are checking token in https://github.com/checkstyle/checkstyle/blob/8cce31164d76c5b4c527a163f6d94a3b4805a316/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java#L143-L147

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColon.java:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        switch(3) {
+            case 3/* space after */ : // violation
```
I am not seeing any good/bad cases with a comment inbetween the `case` and the case's constant. Please add some to show they are handled fine. Comments have a tendency to be in very weird places in the AST tree.

AkMo3 replied:
Done

rnveach commented:
@AkMo3 Isn't this a false positive?
https://akmo3.github.io/NoWhiteSpace/reports/diff/checkstyle/index.html#A9
https://akmo3.github.io/NoWhiteSpace/reports/diff/checkstyle/xref/home/akmo/GitHub/contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyNlowSwitch.java.html#L18

`:` is right on the connection with `)` so there is no whitespace between.

romani commented:
It is false positive, very rare code, but possible.

AkMo3 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java:
```diff
@@ -0,0 +1,220 @@
+
+/**
+ * <p>
+ * Checks that there is no whitespace before the colon in a switch block.
```
What do you think about _**Checks that there is no whitespace before the colon of a case/default.**_

rnveach replied:
Yes, please use.

AkMo3 replied:
Done

AkMo3 commented:
Github, generate report

AkMo3 commented:
> 
> 
> @AkMo3 Isn't this a false positive?
> https://akmo3.github.io/NoWhiteSpace/reports/diff/checkstyle/index.html#A9
> https://akmo3.github.io/NoWhiteSpace/reports/diff/checkstyle/xref/home/akmo/GitHub/contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyNlowSwitch.java.html#L18
> 
> `:` is right on the connection with `)` so there is no whitespace between.

Fixed and added to test case. New report: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8cce311_2021034422/reports/diff/index.html

AkMo3 commented:
@rnveach Please find some time to complete the review
